vnext/Microsoft.ReactNative/Views/FocusZoneViewManager.cpp

```diff
@@ -0,0 +1,151 @@
+
+  auto previousKey = m_isHorizontal ? winrt::system::VirtualKey::Left : winrt::system::VirtualKey::Up;
+  auto nextKey = m_isHorizontal ? winrt::system::VirtualKey::Right : winrt::system::VirtualKey::Down;
+
+  if (e.Key() == previousKey) {
+    auto firstFocusableElement =
+        xaml::Input::FocusManager::FindFirstFocusableElement(senderFrameworkElement).try_as<xaml::DependencyObject>();
+    auto isFirstElementFocused = firstFocusableElement == focusedElement;
+    if (!isFirstElementFocused) {
+      xaml::Input::FocusManager::TryMoveFocus(xaml::Input::FocusNavigationDirection::Previous, findNextElementOptions);
+      e.Handled(true);
+    }
+  } else if (e.Key() == nextKey) {
+    auto lastFocusableElement =
+        xaml::Input::FocusManager::FindLastFocusableElement(senderFrameworkElement).try_as<winrt::DependencyObject>();
+    auto isLastElementFocused = lastFocusableElement == focusedElement;
+    if (!isLastElementFocused) {
+      xaml::Input::FocusManager::TryMoveFocus(xaml::Input::FocusNavigationDirection::Next, findNextElementOptions);
+      e.Handled(true);
+    }
+  } else if (e.Key() == winrt::system::VirtualKey::Tab) {
+    if (!react::uwp::IsXamlIsland()) {
+      auto const &coreWindow = winrt::CoreWindow::GetForCurrentThread();
+      auto isShiftDown = KeyboardHelper::IsModifiedKeyPressed(coreWindow, winrt::system::VirtualKey::Shift);
+
+      winrt::Point anchorTopLeft = winrt::Point(0, 0);
+      xaml::Media::GeneralTransform transform = senderFrameworkElement.TransformToVisual(xamlRoot.Content());
+      winrt::Point anchorTopLeftConverted = transform.TransformPoint(anchorTopLeft);
+      auto exclusionRect = winrt::Rect(
+          anchorTopLeftConverted.X,
+          anchorTopLeftConverted.Y,
+          static_cast<float>(senderFrameworkElement.Width()),
+          static_cast<float>(senderFrameworkElement.Height()));
+      findNextElementOptions.ExclusionRect(exclusionRect);
+
+      auto nextElement = xaml::Input::FocusManager::FindNextElement(
+          isShiftDown ? xaml::Input::FocusNavigationDirection::Up : xaml::Input::FocusNavigationDirection::Down,
+          findNextElementOptions);
+      if (nextElement) {
+        xaml::Input::FocusManager::TryFocusAsync(nextElement, winrt::FocusState::Programmatic);
+        e.Handled(true);
+      }
+    } else {
+      // TODO - need an alternative to winrt::CoreWindow::GetForCurrentThread(); for XAML Islands.
+    }
+  }
+}
+
+void FocusZoneViewShadowNode::OnGettingFocus(
+    const winrt::IInspectable &sender,
+    const xaml::Input::GettingFocusEventArgs &e) {
+  if (e.FocusState() != winrt::FocusState::Programmatic && m_lastElement.get() != nullptr &&
+      e.NewFocusedElement() != m_lastElement.get()) {
+    e.TrySetNewFocusedElement(m_lastElement.get());
+  }
+}
+
+void FocusZoneViewShadowNode::OnLosingFocus(
+    const winrt::IInspectable &sender,
+    const xaml::Input::LosingFocusEventArgs &e) {
+  m_lastElement = e.OldFocusedElement();
```

Comment:
do you need to set `Handled`?

Reply:
Not sure about this one - I'm thinking that there might be actions related to focus visual taken by the system, so I'd pass this one up the routed chain. What do you think?